### PR TITLE
docs(geometry): Convention blocks for bbox/boxes (batch 4c: masks, generators, 3D, video)

### PR DESCRIPTION
**Pull request by ducha-aiki**
## Summary

Documents and pins the sixteen mask, generator, 3D, and video symbols of conventions batch 4c, plus the `Boxes3D` class anchor. Callers can read the size-argument order, the three rounding policies, the inclusive extents, the 3D vertex order, the export-mode defaults, and the `VideoBoxes` storage contract directly in the API docs.

- Add the `Boxes3D` Convention block and anchor pointers on `get_boxes_shape`, `from_tensor`, `to_tensor`, `to_mask`, `transform_boxes`, `transform_boxes_`; own blocks for `Boxes.to_mask`, `VideoBoxes`, and the six `kornia.geometry.bbox` functions `validate_bbox3d`, `infer_bbox_shape3d`, `bbox_to_mask`, `bbox_to_mask3d`, `bbox_generator`, `bbox_generator3d`.
- Correct the documented facts that execution contradicted: `validate_bbox3d` checks equal edge extents, not cubes; `Boxes3D.to_mask` returns `(N, depth, height, width)`, not `(N, depth, width, height)`; `Boxes.to_mask` returns `(N, height, width)`, not `(N, width, height)`; `Boxes3D.from_tensor` validation is "positive in the mode's convention", not ">= 2 for _plus"; both `to_mask` doctest equivalence comments named the wrong mode; `get_boxes_shape` omitted depth; the generators described "the masked image". After review: `VideoBoxes` no longer lists `pad` and `to` as copying methods (they update `self` in place, with `unpad` and `type`); the `Boxes3D` anchor no longer says `validate_bbox3d` requires unbatched input; `bbox_to_mask3d` warns that a box covering or overhanging a full axis fills the whole volume (#4255).
- Add 43 convention/wart test nodes per dtype (85 per two-dtype run), all additive, including strict xfails for the 3D generator extent (#4018), the rank-4 3D helper crash, the `Boxes3D.to_tensor` default-mode split, the `VideoBoxes` inherited-method breakage, and the `bbox_to_mask3d` full-axis fill (#4255). A copy-versus-in-place pin covers the `VideoBoxes` inherited methods. The CUDA/export mask path is pinned against the loop path on every device by patching the module's `is_exporting` predicate.
- Re-executed every audit row that #4112, #4189, #4184, #4218, and #4138 changed since the up-front audit; nothing obsolete is re-documented.

Runtime AST of both `kornia/` files with docstrings stripped is identical to main. No existing test node changed.

Part of #4020; follows merged batches 4a #4060 and 4b #4245. Closes the batch's documentation scope.

## Findings and disposition

| Issue | What this batch documents | Classification / disposition |
|---|---|---|
| #3934 | inclusive `+1` in `infer_bbox_shape3d`, `bbox_generator`, the `Boxes3D` stored form | window item; documented as-is |
| #4009 | `'vertices'` export read one larger by the 3D free functions (3D twin, pinned) | MERGE IN WINDOW |
| #4012 | `Boxes3D` constructor rejects integers, `from_tensor` casts (3D twin, pinned) | MERGE IN WINDOW |
| #4013 | `validate_bbox3d` raises; equal-edge check, not a cuboid check | MERGE IN WINDOW |
| #4014 | `bbox_to_mask(width, height)` vs `Boxes.to_mask(height, width)` | MERGE IN WINDOW |
| #4015 | inclusive raw floats vs round/half-open vs truncate, 2D and 3D | MERGE IN WINDOW |
| #4018 | `bbox_generator3d` extent is `size + 1` | MERGE IN WINDOW; strict xfail for the canonical extent |
| #4248 | rank-4 3D input passes `validate_bbox3d`, then crashes or returns wrong shapes | broken output, idiom 3; n/a — single symbol |
| #4249 | `VideoBoxes.get_boxes_shape`/`to_mask` raise `TypeError`; indexing drops `temporal_channel_size` | broken output, idiom 3; n/a — single symbol |
| #4250 | `bbox_to_mask3d` returns float32 `(B, 1, D, H, W)` regardless of dtype | idiom 2; MERGE IN WINDOW with #4015 |
| #4251 | `Boxes3D.to_tensor()` ignores the stored mode label | 2D/3D split; MERGE IN WINDOW |
| #4252 | list-backed `Boxes.to_mask` fills pixel (0, 0) for padding rows | broken output; n/a — single symbol |
| #4255 | `bbox_to_mask3d` fills the whole volume when a box covers or overhangs a full axis | broken output; fix directly, found in review |

## Validation

- Full `pre-commit run --all-files`: passed.
- Both files on CPU float32/float64: 297 passed, 30 expected strict xfails.
- CPU float16/bfloat16: 285 passed, 30 xfailed, the same 12 failure IDs as freshly measured main.
- MPS float32/float16: 286 passed, 30 xfailed, 8 skipped, the same 3 failure IDs as freshly measured main.
- Changed-module doctests: 15 passed. Sphinx `-W` HTML build: passed; all 18 API entries render their block or pointer.
- Every pin mutation-checked against a targeted source or test inversion; simulated fixes turn the five strict xfails into XPASS, and a `pad` that clones fails the new copy-versus-in-place pin.
- No known-failure manifest edits. CUDA and alternate PyTorch versions were not run.

Executed from the worktree root with verified import resolution, PyTorch 2.14.0.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

### CODE REVIEW

File: kornia/geometry/bbox.py [hunk 1]
```diff
@@ -296,15 +350,33 @@ def bbox_to_mask(boxes: torch.Tensor, width: int, height: int) -> torch.Tensor:
 def bbox_to_mask3d(boxes: torch.Tensor, size: tuple[int, int, int]) -> torch.Tensor:
     """Convert 3D bounding boxes to masks. Covered area is 1. and the remaining is 0.
 
+    Convention:
+        ``size`` is ``(depth, height, width)`` and the mask comes back as :math:`(B, 1, depth, height, width)` in
+        ``float32`` whatever the input dtype, unlike :func:`bbox_to_mask`, which keeps the input dtype and has no
+        channel axis, and :meth:`kornia.geometry.boxes.Boxes3D.to_mask`, which keeps the box dtype and returns
+        :math:`(N, depth, height, width)`. After :func:`validate_bbox3d`, which raises ``AssertionError`` for an
+        invalid box, the bounds are read from fixed vertex positions, truncated toward zero with ``.long()``, and
+        compared inclusively, which reads the vertices as inclusive: pass the ``'vertices_plus'`` export. There
```
Comment (ducha-aiki) [inline kornia/geometry/bbox.py]: [P2] Document and pin the full-axis mask failure

This convention describes inclusive bounds without the existing exception when a box spans an entire output axis. For example, `b = Boxes3D.from_tensor(torch.tensor([[0., 1., 1., 4., 2., 2.]]), 'xyzxyz_plus')` followed by `bbox_to_mask3d(b.data, (4, 4, 5))` fills all 80 voxels, whereas `b.to_mask(4, 4, 5)` fills the intended 20. The union-of-planes intermediate becomes all true when any axis is fully covered, so its reductions lose the other two bounds. This reproduces on the merge base too. Since #4020 requires documenting actual behavior, add a warning and an additive wart pin, and route the runtime repair to a separate issue; the current interior-box tests miss it.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-6).

File: kornia/geometry/boxes.py [hunk 2]
```diff
@@ -1017,6 +1038,28 @@ class VideoBoxes(Boxes):
...
+        - :meth:`from_tensor` stores the :math:`(B, T, N, 4, 2)` input unchanged as batched
+          :math:`(B \cdot T, N, 4, 2)` ``'vertices_plus'`` data; there is no mode argument, no conversion and no
+          validation, and integer input is cast to ``float32``. Any other rank or last dimensions, and list input,
+          raise ``ValueError``.
+        - :meth:`to_tensor` accepts every :class:`Boxes` export mode and restores the temporal axis, so
+          ``to_tensor('xyxy')`` is :math:`(B, T, N, 4)`. Its default is the stored ``'vertices_plus'`` mode.
+        - A transformation matrix must carry one entry per flattened frame, :math:`(B \cdot T, 3, 3)`; a
+          :math:`(3, 3)` matrix raises ``ValueError``. Methods that copy through :meth:`clone`, among them
+          :meth:`transform_boxes`, :meth:`translate`, :meth:`clamp`, :meth:`filter_boxes_by_area`, :meth:`pad`,
+          :meth:`merge` and :meth:`to`, return a :class:`VideoBoxes` with the same
```
Comment (ducha-aiki) [inline kornia/geometry/boxes.py]: [P2] Separate mutating methods from methods that clone

`VideoBoxes` inherits `pad` and `to`, and both mutate and return `self`; neither copies through `clone`. Describing them as copying can cause callers to unexpectedly modify the original video's coordinates or dtype. Split those two out or describe this list only as preserving temporal metadata, and pin their identity/mutation behavior. The added test checks only `temporal_channel_size`, so it passes despite this incorrect copy claim.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-6).

File: kornia/geometry/boxes.py [hunk 3]
```diff
@@ -1099,13 +1146,48 @@ class Boxes3D:
...
+          :meth:`to_tensor` only. Mode strings are lowercased before use.
+        - :meth:`to_tensor` reduces the stored vertices with ``amin``/``amax``, so every export is an axis-aligned
+          bounding box and ``to_tensor('vertices_plus')`` is not the identity on :attr:`data` for a rotated box.
+          Its default mode is ``'xyzxyz'`` whatever the stored label, unlike :meth:`Boxes.to_tensor`, which
+          defaults to the stored mode.
+        - :meth:`get_boxes_shape` returns ``(depths, heights, widths)`` in that order, in the inclusive terms.
+        - :func:`~kornia.geometry.bbox.validate_bbox3d`, :func:`~kornia.geometry.bbox.infer_bbox_shape3d` and
+          :func:`~kornia.geometry.bbox.bbox_to_mask3d` have no mode argument and read their input as inclusive:
+          pass them the ``'vertices_plus'`` export, never ``'vertices'``, which they read as one larger per axis.
+          They also require unbatched :math:`(N, 8, 3)` input; see their warnings.
```
Comment (ducha-aiki) [inline kornia/geometry/boxes.py]: [P3] Exclude the validator from the unbatched-only requirement

The preceding list includes `validate_bbox3d`, but that function accepts `(B, N, 8, 3)` as well as `(N, 8, 3)`. The new validator convention and `test_wart_rank4_input_passes_validate_bbox3d_then_breaks_the_3d_helpers_4248` explicitly establish this. Qualify the unbatched requirement as applying to the shape/mask consumers so this class anchor does not contradict the behavior pinned elsewhere in the PR.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-6).

**Review by ducha-aiki — COMMENTED** [review]
Changes needed for the audit documentation: the new mask contract misses an existing full-axis failure, and two convention statements contradict execution. These findings call for documentation and additive pins, preserving this batch's no-runtime-change scope.

Validation: both changed test modules on CPU float32/float64: 287 passed, 28 xfailed; changed-module doctests: 15 passed. Both library files have identical runtime ASTs to the merge base after stripping docstrings. The full-axis mask reproduction was checked on both the PR and merge base with verified import paths. CUDA, MPS, other PyTorch versions, full pre-commit and Sphinx were not rerun in this review.

Posted as a comment review because the authenticated account owns this PR.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-6).

**Comment by ducha-aiki** [comment]
Round 2 at 327fd8f8 addresses all three items of the review: the `bbox_to_mask3d` full-axis fill is documented, pinned and tracked in the new #4255; the `VideoBoxes` block separates the cloning methods from `pad`, `unpad`, `to` and `type`, with a copy-versus-in-place pin; the `Boxes3D` anchor no longer says `validate_bbox3d` requires unbatched input. The PR description carries the new issue row and the updated counts.

Validation on the worktree with verified import resolution, PyTorch 2.14.0: both files CPU float32/float64 297 passed, 30 xfailed; CPU float16/bfloat16 285 passed, 30 xfailed, the same 12 pre-existing `TransformBoxes` failures as before; MPS float32/float16 286 passed, 30 xfailed, 8 skipped, the same 3 pre-existing failures. Changed-module doctests 15 passed, full `pre-commit run --all-files` passed, Sphinx `-W` HTML build passed with both #4255 links rendered. Runtime AST of both `kornia/` files with docstrings stripped is unchanged from main. Every new pin mutation-checked. CUDA and other PyTorch versions were not run.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

File: kornia/geometry/bbox.py (hunk 1 above)
Comment (ducha-aiki) [reply kornia/geometry/bbox.py]: Confirmed and reproduced on the merge base for all three axes and for a box overhanging one axis: `bbox_to_mask3d` fills all 80 voxels of a `(4, 4, 5)` volume where `Boxes3D.to_mask` fills 20, 16, 16 and 20 on both its loop and export paths. Filed #4255 with the cause (`all_{z,y}(Z | Y | X)` is `X | Zfull | Yfull`, so every condition is `True` once one slab is full) and the disposition "fix directly"; the export path of `Boxes3D.to_mask` already computes the intended `&` of the three slabs.

In 327fd8f8: the `bbox_to_mask3d` Convention block states the intersection is recovered only when the box leaves at least one index uncovered on every axis, the warning describes the full-axis fill and links #4255, and the `Boxes3D.to_mask` warning cross-references it. Pins: `test_wart_bbox_to_mask3d_fills_the_whole_volume_when_a_box_spans_an_axis_4255` parametrized over `full_width`, `full_height`, `full_depth` and `overhang`, each asserting 80 against the method's intersection count and the unaffected interior box, plus the strict xfail `test_convention_bbox_to_mask3d_intersects_the_axis_ranges_for_a_full_axis_box_4255`. Mutation check: replacing the union with the three-way `&` fails all four wart cases and turns the xfail into a strict XPASS.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

File: kornia/geometry/boxes.py (hunk 2 above)
Comment (ducha-aiki) [reply kornia/geometry/boxes.py]: Confirmed by execution: `pad`, `unpad`, `to` and `type` return `self` and mutate the stored data or dtype; `transform_boxes`, `translate`, `clamp`, `filter_boxes_by_area` and `merge` return a new `VideoBoxes` and leave the source untouched.

In 327fd8f8 the bullet is split: the clone list is now `transform_boxes`, `translate`, `clamp`, `filter_boxes_by_area` and `merge`, and a second sentence says `pad`, `unpad`, `to` and `type` update `self` in place and return it, keeping the temporal size but changing the original coordinates or dtype. New pin `test_convention_inherited_methods_split_copies_from_in_place_updates` asserts identity and source-untouched for the copying five, identity plus mutation for `pad`/`unpad` and `to`/`type`, and the temporal size on every result; the existing #4249 wart test's comment no longer claims the in-place methods clone. Mutation check: making `pad` clone fails only the new pin, so the gap you named is now covered.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

File: kornia/geometry/boxes.py (hunk 3 above)
Comment (ducha-aiki) [reply kornia/geometry/boxes.py]: Agreed; the validator accepts rank 3 and rank 4 and the anchor contradicted its own block and the #4248 pins. In 327fd8f8 the sentence reads: the validator also accepts batched `(B, N, 8, 3)` input, but the shape and mask helpers require unbatched `(N, 8, 3)` input; see their warnings. The existing `test_wart_rank4_input_passes_validate_bbox3d_then_breaks_the_3d_helpers_4248` already pins the rank-4 acceptance, so no new pin was needed.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

**Review by ducha-aiki — COMMENTED** [review]
Re-reviewed the update at 327fd8f8. All three findings from my previous review are addressed, and I found no new actionable issues in the revision.

- The full-axis mask defect is documented on both relevant APIs, tracked in #4255, and pinned for all three axes plus overhang, with a strict xfail for the correct intersection.
- VideoBoxes now distinguishes copying methods from in-place updates; the new test checks object identity, source preservation, coordinate changes, dtype changes, and temporal metadata.
- The Boxes3D anchor correctly distinguishes the validator's batched-input acceptance from the shape/mask helpers' requirements.

Validation rerun from the updated worktree with verified import resolution: both test modules on CPU float32/float64: 297 passed, 30 xfailed; changed-module doctests: 15 passed. Both library files retain identical runtime ASTs to the merge base after stripping docstrings. GPU/half-precision runs, full pre-commit, and Sphinx were not rerun for this review.

This addresses my review concerns for the documentation-only audit batch. Posting as a comment review because the authenticated account owns the PR.

Posted on behalf of [@ducha-aiki](https://github.com/ducha-aiki) by Codex (gpt-6).
